Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
+        auto* backend = gtk_printer_get_backend(printer);
+        if (!g_strcmp0(G_OBJECT_TYPE_NAME(backend), "GtkPrintBackendFile")) {
+            data.printer = printer;
+            return TRUE;
+        }
+        return FALSE;
+    }, &data, nullptr, TRUE);
+
+    return data.printer;
+}
+
+static void webkitPrintOperationSendPagesToPrintPortal(WebKitPrintOperation* printOperation, WebFrameProxy* webFrame, GtkPrintSettings* printSettings, GtkPageSetup* pageSetup)
+{
+    auto* priv = printOperation->priv;
+
+    RELEASE_ASSERT(priv->portalProxy);
+    RELEASE_ASSERT(priv->preparePrintResponse && priv->preparePrintResponse->token());
+    RELEASE_ASSERT(priv->preparePrintResponse->result() == WEBKIT_PRINT_OPERATION_RESPONSE_PRINT);
+
+    const char* applicationName = g_get_application_name();
+    GUniquePtr<char> jobName(g_strdup_printf("%s job #%u", applicationName ? applicationName : "WebKit", ++jobNumber));
+    GRefPtr<GtkPrinter> filePrinter = findFilePrinter(printSettings);
+
+    // Print the page into a temporary file
+    GUniqueOutPtr<char> filename;
+    int fd = g_file_open_tmp("webkitgtkprintXXXXXX", &filename.outPtr(), nullptr);
+    close(fd);
+    GUniquePtr<char> uri(g_filename_to_uri(filename.get(), NULL, NULL));
+
+    GRefPtr<GtkPrintSettings> modifiedPrintSettings(gtk_print_settings_copy(printSettings));
+    gtk_print_settings_set(modifiedPrintSettings.get(), GTK_PRINT_SETTINGS_OUTPUT_URI, uri.get());
+
+    priv->printJob = adoptGRef(gtk_print_job_new(jobName.get(), filePrinter.get(), modifiedPrintSettings.get(), pageSetup));
+
+    PrintInfo printInfo(priv->printJob.get(), printOperation->priv->printMode);
+    auto& page = webkitWebViewGetPage(printOperation->priv->webView.get());
+    page.drawPagesForPrinting(webFrame, printInfo, [printOperation = GRefPtr<WebKitPrintOperation>(printOperation)](std::optional<WebCore::SharedMemory::Handle>&& data, WebCore::ResourceError&& error) mutable {
+        auto* priv = printOperation->priv;
+        // When running synchronously, WebPageProxy::printFrame() calls endPrinting().
+        if (priv->printMode == PrintInfo::PrintMode::Async && priv->webView)
+            webkitWebViewGetPage(priv->webView.get()).endPrinting();
+
+        if (!data || !error.isNull()) {
+            if (!error.isNull())
+                webkitPrintOperationFailed(printOperation.get(), WTFMove(error));
+            webkitPrintOperationFinished(printOperation.get());
+            return;
+        }
+
+        priv->data = data->releaseHandle();
+        GUniqueOutPtr<GError> jobError;
+        gtk_print_job_set_source_fd(priv->printJob.get(), priv->data.value(), &jobError.outPtr());
+        if (jobError) {
+            webkitPrintOperationFailed(printOperation.get(), GUniquePtr<GError> { g_error_new_literal(WEBKIT_PRINT_ERROR, WEBKIT_PRINT_ERROR_GENERAL, jobError->message) });
+            webkitPrintOperationFinished(printOperation.get());
+            return;
+        }
```

Comment:
I think there's too much code duplication here with the non-portal codepath. I'm not sure what the cleanest way to avoid it would be, though. Maybe a helper function that returns true to say "continue printing" and false to bail?

Reply:
I'm not sure how much you want me to massage this file in a single PR, but I think what I can do for now is make `webkitPrintOperationFailed()` also call `webkitPrintOperationFinished()`, since all callers of one also call the other.